jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java

```diff
@@ -0,0 +1,459 @@
+            if (_forwardedHeader.equalsIgnoreCase(field.getName()))
+                rfc7239.addValue(field.getValue());
+        
+        if (rfc7239.getFor()!=null)
+            return new HostPortHttpField(rfc7239.getFor()).getHost();
+        
+        return null;
+    }
+    
+    private String getXForwardedFor(Request request)
+    {
+        // Get the right most XForwarded-For for value.
+        // This is the value from the closest proxy and the only one that
+        // can be trusted.
+        String forwarded_for = null;
+        HttpFields httpFields = request.getHttpFields();
+        for (HttpField field : httpFields)
+            if (_forwardedHeader.equalsIgnoreCase(field.getName()))
+                forwarded_for = field.getValue();
+        
+        if (forwarded_for==null || forwarded_for.isEmpty())
+            return null;
+        
+        int comma = forwarded_for.lastIndexOf(',');
+        return (comma>=0)?forwarded_for.substring(comma+1).trim():forwarded_for;
+    }
+
+
+    private final class Remote implements Closeable
+    {
+        private final String _ip;
+        private final int _limit;
+        private final Locker _locker = new Locker();
+        private int _passes;
+        private Deque<CompletableFuture<Closeable>> _queue = new ArrayDeque<>();
+        private final CompletableFuture<Closeable> _passed = CompletableFuture.completedFuture(this);
+        
+        public Remote(String ip, int limit)
+        {
+            _ip=ip;
+            _limit=limit;
+        }
+        
+        public CompletableFuture<Closeable> acquire()
+        {
+            try(Locker.Lock lock = _locker.lock())
+            {
+                // Do we have available passes?
+                if (_passes<_limit)
+                {
+                    // Yes - increment the allocated passes
+                    _passes++;
+                    // return the already completed future
+                    return _passed; // TODO is it OK to share/reuse this?
+                }
+                
+                // No pass available, so queue a new future 
+                CompletableFuture<Closeable> pass = new CompletableFuture<Closeable>();
+                _queue.addLast(pass);
+                return pass;
+            }
+        }
+        
+        @Override
+        public void close() throws IOException
+        {
+            try(Locker.Lock lock = _locker.lock())
+            {
+                // reduce the allocated passes
+                _passes--;
+                while(true)
+                {
+                    // Are there any future passes waiting?
+                    CompletableFuture<Closeable> pass = _queue.pollFirst();
+                    
+                    // No - we are done
+                    if (pass==null)
+                        break;
+                    
+                    // Yes - if we can complete them, we are done
+                    if (pass==null || pass.complete(this))
```

Comment:
`pass==null` was already checked the line above. When is it that you move back to IJ ? :smirk: